synapse/handlers/register.py

```diff
@@ -618,7 +615,7 @@ async def appservice_register(self, user_localpart: str, as_token: str) -> str:
         user_id = user.to_string()
         service = self.store.get_app_service_by_token(as_token)
         if not service:
-            raise AuthError(403, "Invalid application service token.")
+            raise AuthError(401, "Invalid application service token.")
```

Comment:
I don't really know why I had to change this, but the following test was failing because it expected a 401 and got a 403: `tests.rest.client.test_register.RegisterRestServletTestCase.test_POST_appservice_registration_invalid`

I checked the spec, it looks like we're supposed to return 401 in those cases

Reply:
oohkay. So it seems this code was never being called before, because previously the token was checked by the call to `get_appservice_by_req` that you've now removed in synapse/rest/client/register.py.

Maybe we should have this raise an InvalidClientCredentialsError instead, for consistency with the old behaviour and because more-specific exception classes are good?

Reply:
Done